scipy/optimize/minpack.py

```diff
@@ -510,28 +523,35 @@ def curve_fit(f, xdata, ydata, p0=None, sigma=None, absolute_sigma=False,
         values will all be 1 (if the number of parameters for the function
         can be determined using introspection, otherwise a ValueError
         is raised).
-    sigma : None or M-length sequence, optional
-        If not None, the uncertainties in the ydata array. These are used as
-        weights in the least-squares problem
-        i.e. minimising ``np.sum( ((f(xdata, *popt) - ydata) / sigma)**2 )``
-        If None, the uncertainties are assumed to be 1.
+    sigma : None or M-length sequence or MxM array, optional
+        Determines the uncertainty in `ydata`. If we define residuals as
+        ``e = ydata - f(xdata, *popt)``, then
+
+            - If `sigma` is 1-d, it determines standard deviations of residuals `e`
```

Comment:
I guess it's not totally correct to call `sigma` as standard deviations of *residuals*.

How about:
```
Determines the uncertainty in `ydata`. If we define residuals as
``e = ydata - f(xdata, *popt)``, then the interpretation of `sigma`
depends on its number of dimensions:

1-d `sigma` contains values of standard deviations of errors in `ydata`.
The optimized function is ``chisq = sum((e / sigma) ** 2)``.

2-d `sigma` is the covariance matrix of errors in `ydata`.
The optimized function is ``chisq = e.T.dot(inv(sigma)).dot(e)``.
```